internal/application/manager.go

```diff
@@ -521,8 +512,119 @@ func (m *dataManager) loadDevices() error {
+	if err != nil {
+		return err
+	}
+
+	err = m.uploadDevices(data.Devices, overwrite)
+	if err != nil {
+		return err
+	}
+
+	m.recordedData = &recordedData{
+		Events:   data.RecordedEvents,
+		Devices:  utils.SliceToMap(data.Devices, func(d coreDtos.Device) string { return d.Name }),
+		Profiles: utils.SliceToMap(data.Profiles, func(dp coreDtos.DeviceProfile) string { return dp.Name }),
+	}
+
+	m.appSvc.LoggingClient().Debugf("ARR Import: Imported %d events, %d devices and %d device profiles",
+		len(m.recordedData.Events), len(m.recordedData.Devices), len(m.recordedData.Profiles))
+	return nil
+}
+
+func (m *dataManager) uploadDevices(devices []coreDtos.Device, overwrite bool) error {
+	deviceClient := m.appSvc.DeviceClient()
+	for _, device := range devices {
+		_, err := deviceClient.DeviceNameExists(context.Background(), device.Name)
+		if err != nil && err.Code() != http.StatusNotFound {
+			return fmt.Errorf("failed check if device %s exist in system: %w", device.Name, err)
+		}
+
+		if err != nil && err.Code() == http.StatusNotFound {
+			addRequest := requests.NewAddDeviceRequest(device)
+			_, err := deviceClient.Add(context.Background(), []requests.AddDeviceRequest{addRequest})
+			if err != nil {
+				return fmt.Errorf("failed to add device %s to system: %w", device.Name, err)
+			}
+
+			m.appSvc.LoggingClient().Debugf("ARR Import: Added new device %s", device.Name)
+			continue
+		}
+
+		if !overwrite {
+			continue
+		}
+
+		updateDevice := coreDtos.UpdateDevice{
+			Name:           &device.Name,
+			Description:    &device.Description,
+			AdminState:     &device.AdminState,
+			OperatingState: &device.OperatingState,
+			ServiceName:    &device.ServiceName,
+			ProfileName:    &device.ProfileName,
+			Labels:         device.Labels,
+			Location:       device.Location,
+			AutoEvents:     device.AutoEvents,
+			Protocols:      device.Protocols,
+			Tags:           device.Tags,
+			Properties:     device.Properties,
+		}
+		updateRequest := requests.NewUpdateDeviceRequest(updateDevice)
+		_, err = deviceClient.Update(context.Background(), []requests.UpdateDeviceRequest{updateRequest})
+		if err != nil {
+			return fmt.Errorf("failed to update device %s in system: %w", device.Name, err)
+		}
+
+		m.appSvc.LoggingClient().Debugf("ARR Import: Updated existing device %s", device.Name)
+	}
+
+	return nil
+}
+
+func (m *dataManager) uploadProfiles(profiles []coreDtos.DeviceProfile, overwrite bool) error {
+	profileClient := m.appSvc.DeviceProfileClient()
+	for _, profile := range profiles {
+		_, err := profileClient.DeviceProfileByName(context.Background(), profile.Name)
+		if err != nil && err.Code() != http.StatusNotFound {
+			return fmt.Errorf("failed check if profile %s exists in system: %w", profile.Name, err)
+		}
+
+		addRequest := requests.NewDeviceProfileRequest(profile)
+
+		// If got here with an err then the Device Profile doesn't exist in the system, so must add it.
+		if err != nil {
```

Comment:
should this use the same logic as the uploadDevices()?

err != nil && err.Code() == http.StatusNotFound

Reply:
Good question. I'll have to look at the API to see what it returns in this the Not Found case vs. general error. 
I think I did look, but will verify.

Reply:
@brian-intel,  Yes it does return 404. Great catch. I'll submit a fix PR later today. THX!